Add Code output and other updates

microbit-grace opened the pull request:
**TODO**
- [x] MakeCode editor top bar showing what class is being detected
- [x] Remove white parts and full screen all the main bits
- [ ] The output events (serial messages) get queued and then lags... maybe add debounce? (Need testing with older computers) This is particularly obvious with sounds.
  - Predict every 250ms.
  - Try stopping sound in blocks 
- [ ] Issue: 1. Connect 2. Record 3rd class, then reset. 3. Record for 1st class, see no prediction.
- [x] Investigate weird MakeCode editor behaviour where the dropdown box is squished(?!)
- [ ] Update extension to use the newest version of extension (with the colourful dots)

**Changes made**
- Remove audio permission request
- Handle error scenarios for connection
- Automatically download initial MakeCode program to micro:bit
- Swap messages to use serial instead of UART
- Flip camera to use back facing cam
- Add "Code" tab "Edit in MakeCode" button which opens embedded MakeCode
- Show preview code blocks in "Code" tab (if "Edit in MakeCode" button has not been clicked on)

**Links**
- [Initial MakeCode program to micro:bit](https://makecode.microbit.org/_FMY92mAfo23k)
- [Custom TM extension](https://github.com/microbit-grace/tm-extension/)

**Known/possible limitations notes**
- Disconnection of webcam cable is not handled. They would need to reload the page and lose their data.
- Initial connection always downloads a program to the micro:bit. If users want to just connect without overwriting their program and keep their custom program, they cannot. 
- It might not be very obvious that you need to "Download" the code for the behaviour of the micro:bit to change...
- When first recording, may need to press button a couple of times because the first couple of messages may not be sent properly.

microbit-grace committed:
Remove audio permission request

microbit-grace committed:
Connect and download program to micro:bit

Handles connection errors and allow for reconnect
without reloading page.

microbit-grace committed:
Parse serial messages for recording data

microbit-grace committed:
Send commands to microbit

microbit-grace committed:
Add comment

microbit-matt-hillsdon commented on src/microbit/Microbit.js:
```diff
@@ -1,49 +1,96 @@
-                }
+            const decoded = event.data;
+            this.serialBuffer = this.serialBuffer + decoded;
+            if (this.serialBuffer.endsWith("\n")) {
```
Is this guaranteed to be at the end? Or can we have got all of one message and some of the next one?

microbit-grace replied:
Yes, I suppose it may be possible to get all of one message and some of the next. Probably makes sense to execute the commands as they come instead of waiting for the messages to accumulate and look completed. I have updated the logic

microbit-grace committed:
Fix serial buffer logic to execute messages with partial messages

microbit-grace committed:
Rip out bluetooth-related connection logic

microbit-grace committed:
Handle micro:bit disconnect/reconnect

microbit-grace committed:
Flip camera to back facing mode

microbit-grace committed:
Use start/stop recording serial messages

to avoid the issue where a message gets lost and the toggling of recording state gets out of sync with button/logo presses on the micro:bit

microbit-grace commented on src/config.js:
```diff
@@ -58,7 +58,7 @@ let GLOBALS = {
 	numClasses: 3,
 	audioContext: new AudioContext(),
-	isBackFacingCam: false,
+	isBackFacingCam: true,
```
Flipped this to true to stop mirroring

microbit-grace committed:
Add code output tab and edit in makecode btn

microbit-grace committed:
Add embedded MakeCode

When using controller 2, I can show the back button by providing `onBack` callback, but the download button default behaviour does not work (`onDownload` callback would be needed).

When using controller 1, the download button works, but adding `onBack` callback no longer provides a back button.

For now, I have implemented controller 1 since we may customise the top bar anyway.

microbit-grace commented on src/outputs/CodeEditor.js:
```diff
@@ -0,0 +1,96 @@
+      "https://makecode.microbit.org",
+      undefined, // Version.
+      undefined, // Language.
+      1, // Controller.
```
For MakeCode embed, when using controller 2, I can show the back button by providing `onBack` callback, but the download button default behaviour does not work (`onDownload` callback seems to be needed?).

When using controller 1, the download button works, but adding `onBack` callback no longer provides a back button.

For now, I have used controller 1 with a custom top bar with a back button. Reimplementing download would require more work and we may decide to add a top bar anyway.

microbit-matt-hillsdon replied:
I was expecting 2 with download reimplemented ... but mainly so we can be the only people manage the USB connection. Are there adverse consequences to letting MakeCode do the flash? E.g. does it stay connected and read/consume subsequent serial?

microbit-grace replied:
Had a go and yes there is adverse consequences. Will use controller 2, but will keep the top bar in case we want it. I think removing it is quite straight forward.

microbit-grace committed:
Reformat files and hook in custom download flow

with custom progress dialog

microbit-grace committed:
Don't download initial program to micro:bit if it has been downloaded before

microbit-grace committed:
Update initial program to use custom TM extension with a default program

microbit-grace committed:
Load code editor when loading output section

microbit-grace committed:
Setup initial MakeCode project

microbit-grace committed:
Update initial program

microbit-grace committed:
Add code preview to code tab

microbit-grace committed:
Remove code preview once editor has been opened (assuming user has made changes to the code)

microbit-grace committed:
Update editor top bar to show detected class

microbit-grace committed:
Remove content under tool and fullscreen tool

microbit-grace committed:
Predict every 250ms

microbit-matt-hillsdon committed:
Fix rendering of dropdowns in the editor

microbit-matt-hillsdon commented:
Moved current limitations / WIP to internal doc for now. Merging this so folks can see the direction.
